### ISSUE TITLE
how to use WFST for context biasing to improve recognition oov in espnet?

### ISSUE THREAD

**Issue by trangtv57** [espnet/espnet#2872]
hi,
I have a problem with recognition oov in ASR (like proper noun) and it's not appear in training set. I have read about wfst +rnn transducer in paper of google  [streaming end to end speech recognition for mobile device](https://arxiv.org/pdf/1811.06621.pdf) , it introduce term "context biasing", it's using wfst for bias oov of personal for improve WER. And I am not sure but seem espnet don't support this. Is it right?
second question, i see some approach about using CTC + language model, because not sure how wfst work with rnn transducer, Can you point me what is difference of 2 solution, and if wfst work better then normal lm +ctc, how can i combine ctc+wfst.
Thanks you.

**Comment by chenjiasheng** [comment]
This is our approch in Mandarin: 
1, Add a FC layer + CTC on top of your RNN-T encoder to outpt pinyin (phonetic of Mandarin) sequence. Fix other layers and train the ctc loss. Then you get the pinyin probability matrix/lattice.
2, Build a FST model cosuming CTC pinyin lattice and producing characters. 
3, When decoding, choose FST result if any FST path is hit, otherwise choose transducer result.

If you are working on English or Germany, you can use G2P to convert your training text into phonetics and train a phonetic CTC layer.
If you are working on Japanese/Korean, the two languages has similar counterparts as pinyin in Mandarin.

**Comment by sw005320** [comment]
> This is our approch on Mandarin:
> 1, Add a FC layer + CTC on top of your RNN-T encoder to outpt pinyin (phonetic of Mandarin) sequence. Fix other layers and train the ctc loss. Then you get the pinyin probability matrix/lattice.
> 2, Build a FST model cosuming CTC pinyin lattice and producing characters.
> 3, When decoding, choose FST result if any FST path is hit, otherwise choose transducer result.
> 
> If you are working on English or Germany, you can use G2P to convert your training text into phonetics and train a phonetic CTC layer.
> If you are working on Japanese/Korean, the two languages has similar counterparts as pinyin in Mandarin.

I like this solution.
This is quite straightforward.
Did you do this with your own wFST-based decoder or other tools?
We're thinking of doing something similar with k2.

**Comment by kingsword5566** [comment]
That sounds great!
Is there any recipe about WFST decoder in ESPnet?

**Comment by tbright17** [comment]
> This is our approch on Mandarin:
> 1, Add a FC layer + CTC on top of your RNN-T encoder to outpt pinyin (phonetic of Mandarin) sequence. Fix other layers and train the ctc loss. Then you get the pinyin probability matrix/lattice.
> 2, Build a FST model cosuming CTC pinyin lattice and producing characters.
> 3, When decoding, choose FST result if any FST path is hit, otherwise choose transducer result.
> 
> If you are working on English or Germany, you can use G2P to convert your training text into phonetics and train a phonetic CTC layer.
> If you are working on Japanese/Korean, the two languages has similar counterparts as pinyin in Mandarin.

Did you use Essen to do ctc-based WFST decoding?

**Comment by jinggaizi** [comment]
@tbright17 hi,  whats's Essen

**Comment by tbright17** [comment]
> @tbright17 hi, whats's Essen

https://github.com/srvk/eesen

**Comment by jinggaizi** [comment]
thanks, i saw a paper that have a similar work--  https://arxiv.org/abs/2104.03587

**Comment by tbright17** [comment]
Also saw this paper. Some previous studies also use Eesen for WFST based
CTC decoding but don't remember exactly which papers. Maybe K2 is a better
choice later on.

**Comment by tbright17** [comment]
> thanks, i saw a paper that have a similar work-- https://arxiv.org/abs/2104.03587

Also saw this paper. Some previous studies also use Eesen for WFST based CTC decoding but don't remember exactly which papers. Maybe K2 is a better choice later on.

**Comment by trangtv57** [comment]
hi,
does we have any update about recipe or example using wfst (maybe k2 or eesen) for rnn transducer model @sw005320 
sorry for late question because i'm switching to another task and now back to forcus improve our asr solution again.

**Comment by sw005320** [comment]
There is a very good PR about it https://github.com/espnet/espnet/pull/3176
It seems that we can easily use wfst-based systems thanks to k2.
Also, @brianyan918 is working on this direction.

**Comment by brianyan918** [comment]
WFST decoding for RNN transducer is an open problem I'm interested in. I'm not aware of a solution as of right now. EESEN is for decoding CTC with WFST decoder - to do the same style decoding in RNN-T requires some thought on how to deal with the autoregressive nature.

**Comment by trangtv57** [comment]
tks @sw005320 , i will follow PR #3176. 
I am also thinking about how can use k2 with rnn-t like with CTC, i hope this PR can help.
Thanks, if have any update useful, pls update with us.

**Comment by trangtv57** [comment]
I just try see detail of PR #3176, but it's just help for use WFST with output of CTC  not rnn-t (i think, but not sure). So I think we have more research for combine rnn-t with WFST

**Comment by sw005320** [comment]
Sorry, my introduction of #3176 is just for the use of WFST decoder in general. As @brianyan918 said, we need some research and development on how we combine FST and RNN-T.

In addition to some more model level integration like https://github.com/espnet/espnet/issues/2872#issuecomment-773076095, one interesting direction is to perform both FST and RNN-T decoding in parallel and combine two hypothesis, or we can perform a two pass decoding algorithm from FST to RNN-T or vice versa. Maybe we can prepare some CTC-based shared encoder?

**Comment by trangtv57** [comment]
sorry, i miss @chenjiasheng 's solution, i got it, tks you.
